Frameworks/Foundation/NSScanner.mm

```diff
@@ -575,147 +339,249 @@ - (BOOL)scanUpToString:(NSString*)string intoString:(NSString**)intoString {
+    // but WinObjC does not have [NSDecimalNumber decimalNumberWithString:] yet
+    double doubleValue;
+    if (![self scanDouble:&doubleValue]) {
+        return NO;
+    }
 
-/**
- @Status Stub
- @Notes
-*/
-- (BOOL)scanHexDouble:(double*)result {
-    UNIMPLEMENTED();
-    return StubReturn();
-}
+    if (decimalValue) {
+        *decimalValue = [[NSDecimalNumber numberWithDouble:doubleValue] decimalValue];
+    }
 
-/**
- @Status Stub
- @Notes
-*/
-- (BOOL)scanHexFloat:(float*)result {
-    UNIMPLEMENTED();
-    return StubReturn();
+    return YES;
 }
 
 /**
  @Status Interoperable
  @Notes
 */
-- (BOOL)scanUnsignedLongLong:(unsigned long long*)pValue {
-    const char* pScanStart = (char*)[_string UTF8String];
-    char* pScanEnd = nullptr;
+- (BOOL)scanHexDouble:(double*)result {
+    enum {
+        STATE_START,
+        STATE_SIGN,
+        STATE_ZERO,
+        STATE_MANTISSA,
+        STATE_P,
+        STATE_EXPONENT,
+    } state = STATE_START;
+
+    bool mantissaIsNegative = false;
+    bool exponentIsNegative = false;
+    bool scannedPeriod = false;
+    bool hasValue = false;
+
+    double mantissa = 0;
+    int exponent = 0;
+
+    // Once a '.' has been scanned, this /= 16 every character, so that each base-16 decimal digit is appropriately placed
+    // ie: 1.1ae = 1 + (1 * (mantissaDigitPlace /= 16)) + (a * (mantissaDigitPlace /= 16)) + (e * (mantissaDigitPlace /= 16))
+    double mantissaDigitPlace = 1;
+
+    const NSUInteger length = _string.length;
+    NSUInteger i = [self _indexOfNextUnskippedCharacter];
+    for (; i < length; ++i) {
+        unichar unicode = _stringChars[i];
+
+        switch (state) {
+            case STATE_START:
+                if (unicode == '-') {
+                    state = STATE_SIGN;
+                    mantissaIsNegative = true;
+                } else if (unicode == '+') {
+                    state = STATE_SIGN;
+                } else if (unicode == '0') {
+                    state = STATE_ZERO;
+                } else {
+                    return NO;
+                }
+                break;
+
+            case STATE_SIGN:
+                if (unicode == '0') {
+                    state = STATE_ZERO;
+                } else {
+                    return NO;
+                }
+                break;
+
+            case STATE_ZERO:
+                if (unicode == 'x' || unicode == 'X') {
+                    state = STATE_MANTISSA;
+                } else {
+                    return NO;
+                }
+                break;
+
+            case STATE_MANTISSA:
+                if (unicode == '.') {
+                    if (!scannedPeriod) {
+                        scannedPeriod = true;
+                    } else {
+                        goto DONE_SCANNING_HEXDOUBLE; // Stop scanning once a second period has been scanned
+                    }
+
+                } else if (unicode >= '0' && unicode <= '9') {
```

Comment:
you could extract the value of the current character first for digits and hex first and then do the scannedPeriod check.  Better yet, you could have a multiplier value which is 16 before you scan period and 1/16 after you scannedPeriod, then you don't need the if checks at all.  also, isdigit instead of actual comparison here?